first build.ps1

tylerl0706 opened the pull request:
## In scope:

* running `build.ps1` checks for missing tools and runs `Invoke-Build Build`
* running `build.ps1 -Clean` checks for missing tools and runs `Invoke-Build Clean` and then `Invoke-Build Build`
* running `build.ps1 -Test` checks for missing tools and runs `Invoke-Build Build` and then `Invoke-Build Test`
* running `build.ps1 -Bootstrap` informs you of what tools you are missing
    - the dependencies it checks for are:
        - openssl for macOS
        - PowerShellGet
        - InvokeBuild

## Out of scope:

* pulling dotnet sdk installation out of current build script

The script is a bit verbose on purpose so that we can easily add additional dependencies.

Resolves https://github.com/PowerShell/PowerShellEditorServices/issues/586

rkeithhill commented on build.ps1:
```diff
@@ -0,0 +1,104 @@
+
+function needsDotNet451TargetingPack () {
+    # how could we check for this?
+    return $false
```
`Get-CimInstance Win32_Product | Where-Object Name -match '\.NET Framework 4\.5\.1 Multi-Targeting Pack'` 

However, this call is slow - ~30 secs on my laptop.  We might want to do this check only during bootstrap.  Or we find a quicker way to test for this requirement.

tylerl0706 replied:
That sounds good. I have it running on only Bootstrap and also write to the console a small message that says the check wasn't made when on windows and building.

SeeminglyScience replied:
@tylerl0706 part of why it's slow is it verifies and repairs all installed MSI packages, potentially breaking installed applications.  Here's a [kb](https://support.microsoft.com/en-us/help/974524/event-log-message-indicates-that-the-windows-installer-reconfigured-al) on it.  We should definitely find an alternative, maybe searching `HKLM:\SOFTWARE` instead.

tylerl0706 replied:
I did my best to look around for _something_ that could indicate the targeting pack in `HKLM:\SOFTWARE` but I couldn't find anything :(

tylerl0706 replied:
If you have any other hints/pointers, I'm happy to try.

SteveL-MSFT replied:
https://docs.microsoft.com/en-us/dotnet/framework/migration-guide/how-to-determine-which-versions-are-installed#net_b

rkeithhill commented:
> @SeeminglyScience  says - part of why it's slow is it verifies and repairs all installed MSI packages, potentially breaking installed applications. 

That's evil. Yeah, we should find a different way.

tylerl0706 commented:
P.S @rkeithhill + @SeeminglyScience I'm fine leaving the Targeting Pack check out and leaving it in the README.

FWIW, Visual Studio installs this targeting pack so maybe we could settle for checking if they have Visual Studio and if they don't, tell them they may need the targeting pack.

SeeminglyScience commented:
@tylerl0706 I vote to just leave it out.  I haven't actually seen any other build scripts include checking for it presumably for these reasons, it's just a huge pain.  If only we could ditch it entirely and just use the core CLI 😒

tylerl0706 commented:
>  If only we could ditch it entirely and just use the core CLI 😒
I was talking to @joeyaiello about this the other day... It might be possible if the proper shim was in place between .NET Core (or Standard) -> .NET Framework. We'd have to ditch 3/4 support IIRC.

tylerl0706 commented:
So I changed the logic a bit based off of the build.ps1 in the PowerShellStandard repo:
https://github.com/PowerShell/PowerShellStandard/blob/master/build.ps1
and from syncing with @SteveL-MSFT and @JamesWTruher .

The changes are:

`build.ps1 -Clean` will do a clean build... in other words, it will clean and then build.
`build.ps1 -Test` will run the tests using `Invoke-Build Test`

I also changed BootstrapBuildEnv to just Bootstrap because it's less verbose.
